README.md
## Too Many Craftings [tmcraftings]
Copyright (C) 2020 Pixet Bits (mateusmoraisdias3@gmail.com)

Current version: 2.6

<img src= "screenshot.png">

### 2.6 Update
\- Lot of tiny bugs<br/>
\* Changed some textures (pistons and enchanted chest)<br/>
\* Other minor code changes<br/>

### 2.0 Update news

\+ Magic circuits system update<br/>
\+ Gold and Cursed Ruby<br/>
\- Furnaces progress bug<br/>
\* Fixed obsidian mining level<br/>
<br/>
This is the official repository for the Too Many Craftings (tmcraftings) mod for the Minetest game.
Read `readme.txt` for more informations about the mod.

[[Mod Wiki](https://github.com/PixetBits/tmcraftings/blob/master/WIKI.pt.md)]

## License
Code:<br/>
Copyright(c) 2020 PixetBits<br/>
see `license.txt` for details.<br/>

Media (textures):<br/>
Copyright(c) 2020 PixetBits;<br/>
Copyright (C) 2010-2018 Various Minetest developers and contributors*; CC-BY-SA 3.0<br/>
see `license.txt` for details.


CHANGELOG.md
### 2.6
\- Lot of tiny bugs

  * Piston cracking
  * Furnace behaviour
  * Missing recipes
  * Other unlisted bugs
  
\* Changed textures

  * Piston (including nodebox shape)
  * Glue Piston (including nodebox shape)
  * Enchanted Chest
  
\* Other changes in the code (readability, syntax, etc)<br/>

## 2.5
\+ Optimization on wires metadata storage system<br/>
\+ Recovering resources system (slabs to blocks)

## 2.4
\- Furnace refuel bug<br/>
\- Magic Post drop bug<br/>
\* Magic Circuits lit instantly

## 2.3
\- Piston crashing game<br/>
\+ Glue piston works in all directions

## 2.2
\* Fixed Magic Postlight recipe

## 2.1
\+ Missing Anvil recipe<br/>
\* Redundancy between iron and silver in magic items' recipes.<br/>

### 2.0
\+ Magic circuits update (redstonelike)

  * Magic Battery
  * Switch
  * Magic Boiler
  * Solar Panel
  
  * Wires
  * Magic Post
  
  * Disperser
  * Thinker
  * Inverser
  * Magic Gate
  
  * Piston
  * Glue Piston

\+ Gold ore<br/>
\+ Cursed Ruby ore

\* Some textures replaced

  * Magic Battery
  * Magic Furnace
  * Obsidian blocks

\- Furnace  progress bug (default:furnace, tmcraftings:mfurnace)


### 1.5
\- Furnace old gui/old infotext bug

\+ Armor system

  * Iron, tungsten and Titanium pieces/armors
  * Anvil crafting station
  * Forge Hammer
  * Leather

### 1.0
\+ Gui Replacement (inventory, furnace, chest, etc)
\+ Furnace, Chest, and Shelves Replacement.
\+ New dungeon loot system
\+ Bones replaced with Grave block

\+ New Ores
  * Iron
  * Tungsten
  * Titanium
  * Silver
  * Magic Crystal
  
\+ New ore tyer system
\+ Simple Magic System (redstonelike)

\+ New crafting station
  * Compressing Machine
  
\* New storange blocks
  * Metal Crate
  * Enchanted Chest

\* Replaced some textures (Carts and Bucket's mods)


craft_armor.lua
-- Check if is armored --
function is_armored(inv)
    
    return not inv:is_empty('armor') and
    minetest.get_item_group(inv:get_stack('armor', 1):get_name(), 'armor') ~= 0
end

-- Armor rendering --
function update_armorskin(char)

    local inv = char:get_inventory()
    local prp = char:get_properties()
    local skn = prp.textures[1]

    -- Add overlap --
    if is_armored(inv) then

        local a_nm = inv:get_stack('armor', 1):get_name()

        -- Remove other armor --
        if skn:find('%^') then
        
            local idx = skn:find('%^')
            skn = skn:sub(1, idx - 1)
        end

        -- Overlap with armor texture --
        prp.textures[1] =
        skn .. '^' .. a_nm:gsub(':', '_') .. '_skin.png'
        
        -- Update --
        char:set_properties(prp)
    
    -- Remove armor --
    elseif skn:find('%^') then
        
        local idx = skn:find('%^')

        prp.textures[1] = skn:sub(1, idx - 1)
        char:set_properties(prp)
    end
end

minetest.register_on_player_inventory_action(update_armorskin)

-- Armor protection --
minetest.register_on_player_hpchange(function(char, hp_change, reason)
    
    local inv = char:get_inventory()

    if is_armored(inv) then

        local def = inv:get_stack('armor', 1):get_name()
        local crs = reason.type

        def = minetest.get_item_group(def, 'armor') * 10

        -- Valid protection --
        if (crs == 'fall'
        or crs == 'punch'
        or crs == 'node_damage')
        and char:get_hp() > 3 then
            
            char:set_hp((def * math.abs(hp_change)) / 100, 'armor_defense')
        end
    end
end)

-- Hammer durabilty --
minetest.register_on_craft(function(itmstck, char, old_grd)

    local itm = itmstck:get_name()

    if minetest.get_item_group(itm, 'armor') ~= 0 then

        local name = minetest.registered_items[itm].description
        minetest.chat_send_all(char:get_player_name() .. ' forged an ' .. name)

        local inv = char:get_inventory()
        local itm = old_grd[#old_grd]

        itm:add_wear(0.05 * 65535)
        inv:set_stack('craft', 10, itm)
    end
end)

craft_chest.lua
--# Dungeon loot -------------------------------------------#--

    -- Dungeon Loot --
    local loot = {
                    
        'tmcraftings:iron', 'tmcraftings:iron_pick', 'tmcraftings:iron_shovel'  ,
        'default:cobble'  , 'default:coal_lump'    , 'tmcraftings:magic_crystal',
        'default:wood'    , 'default:stick'        , 'default:pick_wood'        ,
        'farming:string'  , 'farming:wheat'        ,
    }

    local found_chest = {}

    -- Replace default function to add an callback --
    core.add_node =
    function(pos, node)
        
        -- Store chests' data --
        if node.name == 'default:chest' then

            -- Active trigger --
            table.insert(found_chest, {

                ['found'] = true,
                ['pos']   = {
                
                    ['x'] = pos.x,
                    ['y'] = pos.y,
                    ['z'] = pos.z,
                }
            })
        end

        minetest.set_node(pos, node)
    end

    function add_loot(item)

        table.insert(loot, item)
    end

    function gen_loot()

        for i, v in pairs(found_chest) do

            if v.found then

                local meta = minetest.get_meta(v.pos)
                local inv = meta:get_inventory()

                meta:set_string('infotext', 'loot')

                -- Disable trigger --
                found_chest[i] = nil
                
                -- Get max slot index --
                local size = inv:get_size('main')

                -- Clear inventory --
                inv:set_list('main', {})

                local val1, val2, val3 = math.abs(v.pos.x), math.abs(v.pos.y), math.abs(v.pos.z)
                math.randomseed(val1 + val2 + val3)

                local itms = {

                    [math.random(0, size)] = loot[val3 % (#loot + 1)],
                    [math.random(0, size)] = loot[val2 % (#loot + 1)],
                    [math.random(0, size)] = loot[val1 % (#loot + 1)],

                    [math.random(0, size)] = loot[(val1 + 1) % (#loot + 1)],
                    [math.random(0, size)] = loot[(val2 + 2) % (#loot + 1)],
                    [math.random(0, size)] = loot[(val3 + 3) % (#loot + 1)],
                }

                for i, v in pairs(itms) do

                    local d_itm = ItemStack({name = v})

                    -- Add more items --
                    if (

                        minetest.get_item_group(v, 'pickaxe') ~= 0 and
                        minetest.get_item_group(v, 'shovel' ) ~= 0 and
                        minetest.get_item_group(v, 'axe'    ) ~= 0 and
                        minetest.get_item_group(v, 'sword'  ) ~= 0

                    ) then d_itm:set_count(math.random(1, 5))
                    
                    -- Fix duplicated tools bug --
                    else
                        
                        d_itm:set_count(1)
                        
                        -- Add random wearing --
                        d_itm:add_wear(math.random(0.20 * 65535, 0.75 * 65535))
                    end

                    inv:set_stack('main', i, d_itm)
                end
            end
        end
    end

    -- Replace loot --
    minetest.register_on_generated(gen_loot)

--#---------------------------------------------------------#--

chests = {}

function can_dig(pos)
    
    local meta = minetest.get_meta(pos)
    local inv  = meta:get_inventory()

    local out = inv:is_empty('main')
    
    if out ~= nil then return out
    else return true end
end

-- Code from original chest --
function can_opn(pos)

    local above = {x = pos.x, y = pos.y + 1, z = pos.z}
    local def = minetest.registered_nodes[minetest.get_node(above).name]

    if def and (

        def.drawtype == "airlike" or
        def.drawtype == "signlike" or
        def.drawtype == "torchlike" or

        (def.drawtype == "nodebox" and def.paramtype2 == "wallmounted")

    ) then return true end
    
    return false
end

-- Based on original code --
function close_chest(sound)

    minetest.sound_play(sound or 'default_chest_close', {

        gain = 0.3,
        pos  = pos,
        max_hear_distance = 10

    }, true)

    for i, v in pairs(chests) do
        
        if v then
            
            minetest.after(0.2, function()
                
                s_node(i, v)
                chests[i] = nil
            end)
        end
    end
end

-- On close gui --
minetest.register_on_player_receive_fields(
function(player, formname, fields)

    if formname == 'tmcraftings:chestformspec' then

        close_chest()
    
    elseif formname:sub(1, 17) == 'tmcraftings:chest' then
        
        close_chest('doors_steel_door_close')
    end
end)

craft_furns.lua
-- Own furnaces behaviours --
function default_timer(pos, elapsed)

    local meta = minetest.get_meta(pos)
    local inv = meta:get_inventory()

    local src = inv:get_list('src')
    local dst = inv:get_list('dst')
    local ful = inv:get_list('fuel')

    local cookitm, outpitm = minetest.get_craft_result({method = "cooking", width = 1, items = src})
    local fuel_tm, outfuel = minetest.get_craft_result({method = "fuel", width = 1, items = ful})

    -- Refuel --
    if meta:get_float('fire') <= 0 and fuel_tm.time ~= 0 then

        meta:set_float('fire', fuel_tm.time)
        meta:set_float('fmax', fuel_tm.time)
        inv:set_stack('fuel', 1, outfuel.items[1])
    end

    -- Turn off --
    if inv:get_stack('src', 1):get_count() == 0 or meta:get_float('fire') <= 0 then
        
        meta:set_string('formspec', get_gui('dfurnace', 0))
        meta:set_float('timer', -1)
    end

    --                                         --
    -- timer == -1 means new cycle or no cycle --
    --                                         --

    -- Successful crafting --
    if cookitm.time ~= 0 then

        local timer = meta:get_float('timer')
        local firet = meta:get_float('fire')

        if timer == -1 and firet > 0 then
            
            meta:set_float('timer', cookitm.time)
            return true
            
        elseif firet > 0 then

            -- Update formspec --
            asize = timer / cookitm.time * 100
            fsize = math.max(firet, 0) / meta:get_float('fmax') * 100

            meta:set_string('formspec', get_gui('dfurnace', 100 - asize, fsize))

            if timer <= 0 then

                inv:set_stack('src', 1, outpitm.items[1])
                inv:add_item('dst', cookitm.item)
                
                meta:set_float('timer', -1)
                meta:set_string('formspec', get_gui('dfurnace', 0, fsize))

                -- Keep burning --
                if inv:get_stack('src', 1):get_count() > 0 then return true end

            else
                
                timer = timer - elapsed
                firet = firet - (elapsed / 5)

                meta:set_float('timer', timer)
                meta:set_float('fire' , firet)

                return true
            end
        
        else return false end
    end
end

function fuelless_timer(pos, elapsed)

    local meta = minetest.get_meta(pos)
    local inv = meta:get_inventory()

    local src = inv:get_list('src')
    local dst = inv:get_list('dst')

    local cookitm, outpitm = minetest.get_craft_result({method = "cooking", width = 1, items = src})

    -- Reset arrow --
    if inv:get_stack('src', 1):get_count() == 0 then
        
        meta:set_string('formspec', get_gui('mfurnace', 0))
        meta:set_float('timer', -1)
    end

    --                                         --
    -- timer == -1 means new cycle or no cycle --
    --                                         --

    -- Successful crafting --
    if cookitm.time ~= 0 then

        local timer = meta:get_float('timer')

        if timer == -1 then
            
            meta:set_float('timer', cookitm.time)
            return true
            
        else

            -- Update formspec --
            asize = timer / cookitm.time * 100
            meta:set_string('formspec', get_gui('mfurnace', 100 - asize))

            if timer <= 0 then

                inv:set_stack('src', 1, outpitm.items[1])
                inv:add_item('dst', cookitm.item)
                
                meta:set_float('timer', -1)
                meta:set_string('formspec', get_gui('mfurnace', 0))

                -- Keep burning --
                if inv:get_stack('src', 1):get_count() > 0 then return true end

            else
                
                timer = timer - elapsed
                meta:set_float('timer', timer)

                return true
            end
        end
    end
end

-- Boiler behaviour --
function on_load_boiler(pos)

    local meta = minetest.get_meta(pos)

    meta:set_string('formspec', get_gui('mgboiler'))
    meta:set_float('fuel', 0)

    local inv = meta:get_inventory()
    inv:set_size('src' , 4)
end

function boiler_behaviour(pos, elapsed)

    local meta = minetest.get_meta(pos)
    local inv  = meta:get_inventory()
    local slots = inv:get_list('src')

    local clit
    for indx, fuel in pairs(slots) do
    
        if meta:get_float('fuel') <= 0 then

            local time, outf = minetest.get_craft_result({

                method = "fuel",
                width  = 1     ,
                items  = {fuel},
            })

            if time then

                time = time.time

                -- Give output --
                inv:set_stack('src', indx, outf.items[1])

                -- Update fuel time --
                clit = meta:get_float('fuel')
                meta:set_float('fuel', clit + time)
                
                return true
            end
        else return true end
    end
end

-- Turn on --
minetest.register_abm({

    nodenames = {

        'tmcraftings:magic_boiler'   ,
        'tmcraftings:magic_boiler_on',
    },

    interval  = 1,
    chance    = 1,
    
    action =
    function(pos, node)
        
        local meta = minetest.get_meta(pos)
        local clit = meta:get_float('fuel')

        meta:set_float('fuel', clit - 1)
        meta:set_string('infotext', 'fuel time left: ' .. (math.max(clit - 1, 0)))

        if clit > 0 then
            
            if node.name ~= 'tmcraftings:magic_boiler_on' then
            
                s_node(pos, 'tmcraftings:magic_boiler_on')
            end

        else
            
            if node.name ~= 'tmcraftings:magic_boiler' then
            
                s_node(pos, 'tmcraftings:magic_boiler')
            end
        end
    end
})

-- Furnace code from default --
function is_diggable(pos)

    local meta = minetest.get_meta(pos)
    local inv = meta:get_inventory()
    
    return inv:is_empty('fuel') and inv:is_empty('dst') and inv:is_empty('src')
end

function can_put_itm(pos, l_nm, indx, stck)

    local meta = minetest.get_meta(pos)
    local inv = meta:get_inventory()
    
    if l_nm == "fuel" then
        
        -- Check if item is burnable --
        if minetest.get_craft_result({
            
            method = "fuel",
            width = 1,
            items = {stck}

        }).time ~= 0 then
        
            return stck:get_count()
            
		else return 0 end
        
    elseif l_nm == "src" then return stck:get_count()
    elseif l_nm == "dst" then return 0 end
end

function can_mov_itm(pos, f_lst, f_idx, tolst, toidx, count)

    local meta = minetest.get_meta(pos)
    local inv  = meta:get_inventory()
    local stck = inv:get_stack(f_lst, f_idx)
    
    return can_put_itm(pos, tolst, toidx, stck)
end

function can_tak_itm(pos, l_nm, indx, stck)

    return stck:get_count()
end

craft_forms.lua
local def_inv =
    [[
    list[current_player;main;0.25,4.0;8,3;8]
    list[current_player;main;0.25,7.5;8,1;0]
    ]]
    .. default.get_hotbar_bg(0.25, 7.5)

-- Change default inventory --
minetest.register_on_joinplayer(function(char)
    
    -- By the way, render armor --
    update_armorskin(char)

    local inv = char:get_inventory()
    
    -- Enchanted chest --
    inv:set_size('chest', 8 * 3)

    -- Helpful inventory --
    inv:set_size('auxi' , 3)
    inv:set_size('armor', 1)

    -- Anvil slot --
    inv:set_size('craft', 10)

    if not minetest.settings:get_bool('creative_mode') then

        char:set_inventory_formspec([[
            size[8.5,8.5]

            list[current_player;auxi;0.75,0.5;1,3;0]
            image[1.75,1.5;1,1;sfinv_crafting_arrow.png]

            image[0.75,0.5;1,1;gui_hb_bg.png]
            image[0.75,1.5;1,1;gui_hb_bg.png]
            image[0.75,2.5;1,1;gui_hb_bg.png]


            list[context;craft;2.75,0.5;3,3]
            list[current_player;craftpreview;6.75,1.5;1,1]
            image[5.75,1.5;1,1;sfinv_crafting_arrow.png]
            ]] 
            .. def_inv)
    end
end)

minetest.register_on_dieplayer(function(player, reason)

    if not minetest.settings:get_bool('creative_mode')
    and player:is_player() then

        player:set_inventory_formspec([[
            size[8.5,8.5]

            list[current_player;auxi;0.75,0.5;1,3;0]
            image[1.75,1.5;1,1;sfinv_crafting_arrow.png]

            image[0.75,0.5;1,1;gui_hb_bg.png]
            image[0.75,1.5;1,1;gui_hb_bg.png]
            image[0.75,2.5;1,1;gui_hb_bg.png]


            list[context;craft;2.75,0.5;3,3]
            list[current_player;craftpreview;6.75,1.5;1,1]
            image[5.75,1.5;1,1;sfinv_crafting_arrow.png]
            ]] 
            .. def_inv)
    end
end)

function get_gui(name, item_pc, fuel_pc, data)

    local arrw = item_pc or 0
    local fire = fuel_pc or 0

    local dfr_gui =
        
        [[
        size[8.5,8.5]

        list[current_player;auxi;0.75,0.5;1,3;0]
        image[0.75,0.5;1,1;gui_hb_bg.png]
        image[0.75,1.5;1,1;gui_hb_bg.png]
        image[0.75,2.5;1,1;gui_hb_bg.png]

        list[context;src;2.75,1.0;1,1;]
        list[context;dst;4.75,1.0;2,2;]
        list[context;fuel;2.75,2.0;1,1]
        ]]

        .. "image[3.75,1.0;1,1;gui_furnace_arrow_bg.png^[lowpart:"
        .. (arrw) .. ":gui_furnace_arrow_fg.png^[transformR270]"
        
        .. "image[3.75,2.0;1,1;default_furnace_fire_bg.png^[lowpart:"
        .. (fire) .. ":default_furnace_fire_fg.png]"

        .. def_inv

    local mfr_gui =
        
        [[
        size[8.5,8.5]

        list[current_player;auxi;0.75,0.5;1,3;0]
        image[0.75,0.5;1,1;gui_hb_bg.png]
        image[0.75,1.5;1,1;gui_hb_bg.png]
        image[0.75,2.5;1,1;gui_hb_bg.png]

        list[context;src;2.75,1.5;1,1]
        list[context;dst;4.75,1.5;1,1]
        ]]

        .. "image[3.75,1.5;1,1;gui_furnace_arrow_bg.png^[lowpart:"
        .. (arrw) .. ":gui_furnace_arrow_fg.png^[transformR270]"
        .. def_inv

    local blr_gui =
        
        [[
        size[8.5,8.5]

        list[current_player;auxi;0.75,0.5;1,3;0]
        list[context;src;3.25,1.0;2,2]
        ]]
        .. def_inv
        
    local cht_gui =
        
        [[
        size[8.5,8.5]
        ]]

        .. "list[" .. (data or 'context;main') .. ";0.25,0.25;8,3;]"
        .. def_inv

    local anv_gui =

        [[
        size[8.5,9.0]

        list[current_player;craft;2.75,0.25;3,3]
        
        list[current_player;craft;2.5,3.5;1,1;9]
        image[2.5,3.5;1,1;tmcraftings_hammer_icon.png]

        list[current_player;craftpreview;3.75,3.5;1,1]

        list[current_player;armor;5.0,3.5;1,3;0]
        image[5.0,3.5;1,1;tmcraftings_shield_icon.png]

        list[current_player;main;0.25,4.75;8,3;8]
        list[current_player;main;0.25,8.0;8,1;0]
        ]]
        .. default.get_hotbar_bg(0.25, 8.0)

    if name == 'dfurnace' then return dfr_gui end
    if name == 'mfurnace' then return mfr_gui end
    if name == 'mgboiler' then return blr_gui end
    if name == 'defchest' then return cht_gui end
    if name == 'defanvil' then return anv_gui end
end